README.md

```diff
@@ -29,122 +37,1058 @@ The library design is inspired by the work in [Exploring HTN Planners through ex
+    a1f7280:::selected
+    a1f7280 --> stop
+    classDef error stroke:#f00
+    classDef selected stroke:#0f0
+```
+
+Here we see here that the planner needs to perform 3 iterations (the diamond nodes) to find a plan to the target.
+
+Methods can also call other methods. We can now write `plusThree` using `plusOne` and `plusTwo`
+
+```typescript
+const plusThree = Task.from<number>({
+	condition: (state, { target }) => target - state > 2,
+	// Methods can be referenced from other methods
+	method: (_, { target }) => [plusTwo({ target }), plusOne({ target })],
+	description: '+3',
+});
+```
+
+Methods, however cannot call themselves recursively or as part of call chain. This is forbidden to prevent infinite recursion. Using recursion means the method will effectively never be called.
+
+## Lenses
+
+Let's say now that we want our agent to manage multiple counters. Let's redefine our system state once more to do this
+
+```typescript
+type System = { counters: { [key: string]: number } };
+```
+
+We can now easily (?) redefine our `plusOne` task to handle this case.
+
+```typescript
+const plusOne = Task.from<System>({
+	// This task will be chosen only if one of the keys is smaller than the target
+	condition: (state, { target }) =>
+		Object.keys(state.counters).some(
+			(k) => state.counters[k] < target.counters[k],
+		),
+	effect: (state, { target }) => {
+		// We find the first counter below the target, we know it exists because of
+		// the condition so we use the non-null assertion (!) at the end
+		const key = Object.keys(state._.counters).find(
+			(k) => state._.counters[k] < target.counters[k],
+		)!;
+
+		// Update the changed counter
+		state._.counters[key]++;
+	},
+	description: '+1',
+});
+```
+
+The above code achieves the same as the single counter `plusOne` (increase one of the counters), but is a bit convoluted, and we cannot tell from the description which counter will be increased. We can improve this setting the `lens` property when defining the action.
+
+```typescript
+const plusOne = Task.of<System>().from({
+	// The lens defines the part of the state where this task is applicable
+	// this property defaults to '/', meaning the root object
+	lens: '/counters/:counterId',
+	condition: (counter, { target }) => counter < target,
+	effect: (counter) => ++counter._,
+	description: ({ counterId }) => `${counterId} + 1`,
+});
+```
+
+We are again introducing a few new things here, but hopefully the code above is relatively intuitive. This declares `plusOne` as a task on the `System` type, but that acts on a specific counter, specified by the lens property. Since we don't know a priori which counter this will apply to, we replace the name of the counter with the placeholder variable, `:counterId`. This variable is passed as part of the context to the task constructor properties, we can see this in the use of the value by the description which now we see that it can also be a function acting on the context.
```

Comment:
Too many concepts in that last sentence, IMO. Here's my attempt to split them into separate sentences:

```suggestion
We are again introducing a few new things here, but hopefully the code above is relatively intuitive. This declares `plusOne` as a task on the `System` type, but that acts on a specific counter, specified by the lens property. Since we don't know a priori which counter this will apply to, we replace the name of the counter with the placeholder variable, `:counterId`. This variable is passed as part of the context to the task constructor properties. We can see this in the definition of `description`, where we use `counterID`.  By the way, this also shows that the description can also be a function acting on the context.
```